src/data/finite/card.lean

```diff
@@ -92,6 +92,38 @@ by { haveI := fintype.of_finite α, haveI := fintype.of_surjective f hf,
 lemma card_eq_zero_iff [finite α] : nat.card α = 0 ↔ is_empty α :=
 by { haveI := fintype.of_finite α, simp [fintype.card_eq_zero_iff] }
 
+lemma card_le_of_injective' (f : α → β) (hf : function.injective f)
+  (h : nat.card β = 0 → nat.card α = 0) : nat.card α ≤ nat.card β :=
+(or_not_of_imp h).cases_on (λ h, le_of_eq_of_le h zero_le')
+  (λ h, @card_le_of_injective α β (nat.finite_of_card_ne_zero h) f hf)
+
+lemma card_le_of_embedding' (f : α ↪ β) (h : nat.card β = 0 → nat.card α = 0) :
+  nat.card α ≤ nat.card β :=
+card_le_of_injective' f f.2 h
+
+lemma card_le_of_surjective' (f : α → β) (hf : function.surjective f)
+  (h : nat.card α = 0 → nat.card β = 0) : nat.card β ≤ nat.card α :=
+(or_not_of_imp h).cases_on (λ h, le_of_eq_of_le h zero_le')
+  (λ h, @card_le_of_surjective α β (nat.finite_of_card_ne_zero h) f hf)
+
+lemma card_eq_zero_of_surjective (f : α → β) (hf : function.surjective f) (h : nat.card β = 0) :
+  nat.card α = 0 :=
+begin
+  casesI finite_or_infinite β,
+  { haveI := card_eq_zero_iff.mp h,
+    haveI := function.is_empty f,
+    exact nat.card_of_is_empty },
+  { haveI := infinite.of_surjective f hf,
+    exact nat.card_eq_zero_of_infinite },
+end
+
+lemma card_eq_zero_of_injective [nonempty α] (f : α → β) (hf : function.injective f)
+  (h : nat.card α = 0) : nat.card β = 0 :=
+card_eq_zero_of_surjective (function.inv_fun f) (function.inv_fun_surjective hf) h
+
+lemma card_eq_zero_of_embedding [nonempty α] (f : α ↪ β) (h : nat.card α = 0) : nat.card β = 0 :=
```

Comment:
Maybe all these lemmas, but certainly these last two, could use a docstring explaining why this is not concerning. These lemmas in particular are all about the wacky edge case of infinite sets whose cardinality is defined to be `0`.
```suggestion
/-- NB: `nat.card` is defined to be `0` for infinite types. -/
lemma card_eq_zero_of_injective [nonempty α] (f : α → β) (hf : function.injective f)
  (h : nat.card α = 0) : nat.card β = 0 :=
card_eq_zero_of_surjective (function.inv_fun f) (function.inv_fun_surjective hf) h

/-- NB: `nat.card` is defined to be `0` for infinite types. -/
lemma card_eq_zero_of_embedding [nonempty α] (f : α ↪ β) (h : nat.card α = 0) : nat.card β = 0 :=
```